azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -184,38 +186,42 @@ static ExecutableFlow getFlowToRestart(final ExecutableFlow flow,
       final Status originalStatus){
     final ExecutionOptions options = flow.getExecutionOptions();
     if (options == null) {
-      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has ExecutionOptions == null");
+      logger.debug("ExecutableFlow: " + flow.getExecutionId() + " has ExecutionOptions == null");
       return null;
     }
-    final Map<String, String> flowParams = options.getFlowParameters();
-    if (flowParams == null || flowParams.isEmpty()) {
-      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has ExecutionOptions == null");
-      return null;
-    }
-
-    // flow can only retry if not reach the limit
-    final int flowMaxRetryLimit = Integer.parseInt(
-        flowParams.getOrDefault(FlowParameters.FLOW_PARAM_MAX_RETRIES, "0"));
-    if (flowMaxRetryLimit <= 0) {
-      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has exceed retry limit, count = "
-          + flowMaxRetryLimit);
-      return null;
-    }
-    // reduce the count
-    flowParams.put(FlowParameters.FLOW_PARAM_MAX_RETRIES, String.valueOf(flowMaxRetryLimit - 1));
-
     // If the original execution status is restartable non-terminal status, it can be retried
     if (RESTARTABLE_NON_TERMINAL_STATUSES.contains(originalStatus)) {
-      logger.info("Submitted flow for restart: " + flow.getExecutionId());
+      if (flow.getSystemRetriedTimes() >= DEFAULT_SYSTEM_FLOW_RETRY_LIMIT){
+        logger.info("ExecutableFlow: " + flow.getExecutionId() + " has reached max retry limit "
+            + "for non-terminal status, ");
+        return null;
+      }
+      logger.info("Submitting flow for restart: " + flow.getExecutionId());
+      flow.setSystemRetriedTimes(flow.getSystemRetriedTimes() + 1);
```

Comment:
1. use `SystemRetriedTimes` to control retry on `NON_TERMINAL_STATUSES`
2. update `SystemRetriedTimes++` if permitted